pallets/disputes/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use crate::traits::DisputeHooks;
+    use codec::{FullCodec, FullEncode};
+    use frame_support::{
+        dispatch::fmt::Debug, pallet_prelude::*, weights::Weight, BoundedBTreeMap,
+    };
+    use frame_system::pallet_prelude::*;
+    use sp_runtime::traits::{AtLeast32BitUnsigned, Saturating};
+
+    pub(crate) type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// Overarching event type.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+        /// The weights generated by the benchmarks.
+        type WeightInfo: WeightInfoT;
+        /// The key that links the dispute to the project.
+        type DisputeKey: AtLeast32BitUnsigned
+            + FullEncode
+            + FullCodec
+            + MaxEncodedLen
+            + TypeInfo
+            + Debug
+            + Copy;
+        /// Used to specify the milestones the dispute is being raised on.
+        type SpecificId: AtLeast32BitUnsigned
+            + FullEncode
+            + FullCodec
+            + MaxEncodedLen
+            + TypeInfo
+            + Debug
+            + Copy;
+        /// This is the max length for specifying the reason while raising the dispute
+        type MaxReasonLength: Get<u32>;
+        /// This is number of juries that can be assigned to a given dispute
+        type MaxJurySize: Get<u32>;
+        /// This is number of specifics that can be assigned to a given dispute
+        type MaxSpecifics: Get<u32>;
+        /// The amount of time a dispute takes to finalise.
+        type VotingTimeLimit: Get<<Self as frame_system::Config>::BlockNumber>;
+        /// The origin used to force cancel and pass disputes.
+        type ForceOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+        /// External hooks to handle the completion of a dispute.
+        type DisputeHooks: DisputeHooks<Self::DisputeKey>;
+    }
+
+    /// Used to store the disputes that is being raised, given the dispute key it returns the Dispute
+    /// Key: DisputeKey
+    /// Value: Dispute<T>
+    #[pallet::storage]
+    #[pallet::getter(fn disputes)]
+    pub type Disputes<T: Config> =
+        StorageMap<_, Blake2_128Concat, T::DisputeKey, Dispute<T>, OptionQuery>;
+
+    /// Stores the dispute keys that will finalise on a given block.
+    /// Key: BlockNumber
+    /// Value: Vec<DisputeKey>
+    #[pallet::storage]
+    pub type DisputesFinaliseOn<T: Config> = StorageMap<
+        _,
+        Blake2_128Concat,
+        BlockNumberFor<T>,
+        BoundedVec<T::DisputeKey, ConstU32<1000>>,
+        ValueQuery,
+    >;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// A dispute has been raised.
+        DisputeRaised {
```

Comment:
none of these event have been asserted, please do so. 
Use `frame_system::Pallet::<T>::assert_last_event()` for this instead of making your own helper

Reply:
Most of them has been already asserted, will cover the one which are missed